### PR TITLE
cc-wrapper: fix on darwin

### LINKED ISSUE #6447: clang calls readlink on PATH
A tool like GHC will stash away a full path to clang to use for, eg, linking. But when this invocation to clang occurs outside of nix-shell, I get error messages about an invalid flag, `-f`, being passed to `readlink`. In `nix-shell`, everything is fine. So it seems as though a `clang` from the nix store is being invoked, but it, in turn, is invoking `readlink` on PATH which, on darwin, is different from the one in the nix `stdenv`.

### PR DESCRIPTION
**Pull request by acowley**
The ld-wrapper.sh script calls `readlink` in some circumstances. We need
to ensure that this is the `readlink` from the `coreutils` package so
that flag support is as expected.

Without doing this, the following happens with a trivial `main.c`:

```
nix-env -f "<nixpkgs>" -iA pkgs.clang
$ clang main.c -L /nix/../nix/store/2ankvagznq062x1gifpxwkk7fp3xwy63-xnu-2422.115.4/Library -o a.out
readlink: illegal option -- f
usage: readlink [-n] [file ...]
```

The key element is the `..` in the path supplied to the linker via a
`-L` flag. With this patch, the above invocation works correctly on
darwin, whose native `/usr/bin/readlink` does not support the `-f` flag.

Fixes #6447

### CODE REVIEW

**Comment by copumpkin** [comment]
Ah, so the use case here is for interactive invocations of `clang`/`ld`, rather than via Nix scripts, right?

**Comment by acowley** [comment]
Yes, that is correct!

**Comment by copumpkin** [comment]
Given that it's a mass rebuild, this should probably go into `staging` rather than `master`.

**Comment by acowley** [comment]
Ah, okay. Can you move it or shall I?

**Comment by copumpkin** [comment]
I don't think GitHub allows one to repoint a PR. Someone could just merge the patch manually to `staging`, but let's get some feedback first. As I mentioned in the linked issue, this will resolve the particular instance but I'm wondering if there are more robust ways to do this more broadly that might apply here too.

**Comment by acowley** [comment]
Yes, what you said sounded reasonable. I figured that since nothing has happened for several months, fixing this instance now is expedient, and perhaps the lack of broader outcry indicates the problem doesn't arise as frequently as it might. It's a pretty substantial nuisance, as it makes development tools installed via nix much less useful than they otherwise ought to be.

**Comment by vcunat** [comment]
There were multiple impure references in cc-wrapper scripts. I wonder if it's better to hardcode the paths directly 093f58b2c8c4854886acb694081ce90d1e1b2045 or just specify `PATH='@coreutils@/bin'` at the beginning of a script. (IIRC that shouldn't affect `$PATH` value *exported* by the script, but I'm not certain.)

**Comment by copumpkin** [comment]
Setting `PATH` explicitly (and not including its previous value) at the top of our scripts is actually pretty appealing. Advantages (if we make it a policy):
* We don't accidentally inherit stuff from the user's shell
* We don't have to adorn every invocation of a utility with its full path
* We get a loud failure if we forget a dependency, because the utility won't be found
* We could in theory verify semi-automatically that our scripts follow the policy (by looking for `PATH=` in the first few lines, and ensuring that `PATH` isn't referenced on the RHS)

:+1: from me on that approach

**Comment by acowley** [comment]
That sounds like the right approach.

**Comment by acowley** [comment]
This is actually a bit trickier than I anticipated. First, the straightforward idea of just using, e.g., `@coreutils@` in the shell scripts to work with `substituteAll` doesn't work as `cc-wrapper` has a `nativeTools` parameter that can cause that variable to not be defined. So I tweaked that a bit as you can see in the new patches, but it still doesn't work; here's the error I get after a while

```
building path(s) ‘/nix/store/i8lvwn55nbynswx6wdsvw4bvj8aw0scg-stdenv-darwin’
output ‘/nix/store/i8lvwn55nbynswx6wdsvw4bvj8aw0scg-stdenv-darwin’ is not allowed to refer to path ‘/nix/store/7l8l3sp3wk0idx0191qhn4hvdvx0wznq-bootstrap-tools’
error: build of ‘/nix/store/fxsahfv1352licgp95mhwgip109cvj27-stdenv-darwin.drv’ failed
```

Perhaps one of the more experienced darwin folks (e.g. @copumpkin @shlevy @gridaphobe @pikajude ) might suggest a fix. I can work at it, but it's a slog.

**Comment by shlevy** [comment]
The problem is that the final version of cc-wrapper is referencing the bootstrap coreutils, not the final version. One potential fix would be to rewrap the final clang after building the final coreutils.

But is there any reason we need to use identical wrappers for builds and for users? We already have `bashInteractive` for example.

**Comment by acowley** [comment]
I'm not familiar enough with the bring-up of the whole environment to make architectural changes. I had hoped that the `nativeTools` flag would sufficiently twiddle the configuration between bootstrapping and final builds to avoid issues like this. It seems like adding a `ccInteractive` would be a fairly intrusive change if it would be necessary to refer to it when building the stdenv.

**Comment by shlevy** [comment]
The whole point is that it wouldn't be necessary to refer to it when building stdenv, right? Since cc-wrapper works fine in nix builds, it only fails for the user?

**Comment by acowley** [comment]
Oh, I'll have to re-run the test I guess. The build error made me think my attempted rebuild of things broke while building stdenv. These builds take a long time to get back to a development environment I can work with, so I usually kick off a downstream build and check back in a day to read the error messages. I'll try again some other time and only build one thing at a time as that error message is apparently pretty confusing.

**Comment by shlevy** [comment]
You'll have to revert your change. The build error is because the final stdenv contains a reference to the bootstrap tools. If you have a separate ccInteractive, it will refer to the newly-built coreutils, not the bootstrap coreutils.

**Comment by acowley** [comment]
The linked issue for this PR was the problem being exposed in GHC through its use of the linker. So, if I'm understanding the suggestion, it would be to leave `cc-wrapper` as it was, but change all downstream dev tools to use a different `ccInteractive` package?

**Comment by shlevy** [comment]
Ah, yeah. I'm not sure what the best solution is there really.

Hmm, I wonder if the better solution is to include coreutils in nix-env environments by default?

**Comment by acowley** [comment]
Part of my question that I perhaps didn't make suitably clear is why I'm running into this problem with this change: I don't understand what this change does that introduces the illegal stdenv reference. Perhaps that could suggest an easy way to adjust this approach of explicitly setting PATH as it seems like rather a good idea in general.

It seems like the right fix should be adjacent to the current state of the PR given that my original attempt at just using an explicit '@coreutils@` path in the `readlink` call addressed the motivating symptom without any noticeable adverse consequences.

**Comment by shlevy** [comment]
Hmm, are you sure that your original attempt worked? Based on a modern nixpkgs? The `allowed-references` stuff is somewhat recent.

**Comment by acowley** [comment]
Yes, I'm sure. I haven't come across `allowed-references`, though, so perhaps we're still not on the same page.

**Comment by shlevy** [comment]
Sorry, it's `allowedReferences` and `allowedRequisites`. The error message above comes from https://github.com/NixOS/nixpkgs/blob/master/pkgs/stdenv/pure-darwin/default.nix#L296-L302

**Comment by acowley** [comment]
Oh, right, I should have recognized that. I'm actually not sure which specific reference in `stdenv` is causing the trouble. I've ag'ed through the `stdenv` store path in the error message, but there's no `bootstrap-tools` path there, so I'm guessing it's a transitive dep but I'm not sure of an easy(-ish) way of tracking it down.

**Comment by shlevy** [comment]
You can do `nix-store -q --tree` to see what's bringing in the reference to bootstrap-tools

**Comment by acowley** [comment]
Well, that's quite a large tree, but the `bootstrap-tools` references are all over the place: `bash`, `bison`, `perl`, etc. I feel like this is related to the `stageFun` definition that sets `dontPatchShebangs=1` to avoid retaining dependencies. The logic there does a path comparison with the `bootstrapTools` derivation to control the patching behavior. That seems super fragile, so I'm not sure how to leverage it to fix other path substitution issues. It seems like hack-upon-hack, but one option might be to use that same check to cripple the `@coreutils@` substitution.

**Comment by acowley** [comment]
I misunderstood the suggestion to look at`nix-store -q` output. Without this PR, `bootstrap-tools` is still in the `nix-store -q --tree` output for things like `bash`, so I don't know what to look at for the build that errors out. I've pasted the tree [here](http://lpaste.net/147978), so if anyone can identify the problem, please share. All I know how to do is twiddle things and rebuild, which takes a long time.

**Comment by copumpkin** [comment]
You can also get something far clearer with `--graph`, since `--tree` duplicates common parts of the graph and makes the whole thing look far larger than it is. You can just feed `--graph` into graphviz :smile:

**Comment by shlevy** [comment]
@acowley You looked at the graph f or the `.drv` file. We need to see the graph for the build output, i.e. the runtime dependencies graph.

**Comment by acowley** [comment]
I looked at the `.drv` because,

```
$ nix-store -q --tree /nix/store/i8lvwn55nbynswx6wdsvw4bvj8aw0scg-stdenv-darwin
/nix/store/i8lvwn55nbynswx6wdsvw4bvj8aw0scg-stdenv-darwin
error: path ‘/nix/store/i8lvwn55nbynswx6wdsvw4bvj8aw0scg-stdenv-darwin’ is not valid
```

I can read the `out` paths from the mentioned derivations and query them, but is there a faster way of doing so?

**Comment by shlevy** [comment]
Ugh, right, because the build failed. Can you check the dependency tree of the cc-wrapper that's closest to stdenv?

**Comment by shlevy** [comment]
@edolstra How difficult would it be to have nix store the dependency info of a build that failed due to disallowed dependencies (either because of allowedReferences or circular reference in multiple outputs) somewhere for investigations like this?

**Comment by acowley** [comment]
I don't quite follow what you mean by "closest". I looked at the `out` path of `/nix/store/9zf63nb5x5dy7lznf2xdfh5ba0fdki8g-clang-wrapper-3.6.2.drv` (mentioned in the `stdenv` derivation's tree), and `bootstrap-tools` didn't show up there, but I"m not sure if that's the same thing.

**Comment by vcunat** [comment]
> How difficult would it be to have nix store the dependency info of a build that failed due to disallowed dependencies (either because of allowedReferences or circular reference in multiple outputs) somewhere for investigations like this?

The output paths of the failed build remain in the /nix/store (at least for me). You can `grep -R` through them...

**Comment by acowley** [comment]
Small update: I tried taking out the PR lines that clear `PATH` in `gnatlink-wrapper.sh`, `ld-solaris-wrapper.sh` (probably not relevant), and `setup-hook.sh`. This let me get much further. However, I run into an error building the `gnum4` package (needed for building GHC). My suspicion is that I was unsetting a `PATH` that was being relied on by something like `patchShebangs`, and the bootstrapTools shell was leaking in as per the warning in the pure-darwin stdenv definition, but that's just a guess.

I dug into the m4 build failure a bit since it's a smaller thing to look at, and it's kind of interesting. The build failures are due to various checks in m4's build system to accommodate Apple's SDK. Since we still pull a version of LibSystem in, we run into trouble with things like redefinitions of `sigismember` and `sigaction` (from `signal.h` in `m4`'s `lib` directory). I tweaked the C pre-processor guards around a few things to nudge it further along, but then ran into a problem with clock resolution. I stopped digging at this point, but can keep going if anyone thinks it's helpful.

Alternatively, if I take the unpacked nixpkgs m4 package, and `./configure && make` in a regular OS X shell, it builds just fine with the normal OS X CLTools.

So either explicitly setting the `PATH` as suggested is still doing unwanted violence, or we previously had unintended dependencies on the PATH *not* being pruned down so extensively.

To summarize the current situation, the only working version of this PR -- "working" in that the parts of nixpkgs I use all build, and compilers (clang and ghc) can be used outside of a nix shell -- is the one where only the call to `readlink` is changed to `@coreutils@/bin/readlink`. Every other variation of this fix has run aground somewhere.

**Comment by acowley** [comment]
I finally got this working. The last holdup was that `cc-wrapper.sh` also calls `grep` which we should have been ensuring was the `grep` from nix.

I've only built this on darwin so far (and that was before the last rebase, so I'll build again tonight), so other stdenvs likely need adjustment to satisfy the explicit `gnugrep` dependency. Does this approach seem okay to folks?

**Comment by vcunat** [comment]
Ah, right, my assumption in https://github.com/NixOS/nixpkgs/pull/11798#issuecomment-165705219 was wrong &ndash; apparently it *is* necessary to do some kind of save-restore dance. While in these cases it's not too bad due to only having a single external `exec` at the end, it will probably be too cumbersome as a general technique (at least with `bash`).

I assume you have tested this well on darwin? Linux doesn't evaluate ATM, but I'll have a look into it.

#### How bash's environment seems to work
If one makes a change to a variable that is exported, that change *is* propagated to sub-processes. Even if one marks it `local`, commands called *within* the function inherit that (new) value.

**Comment by acowley** [comment]
Yes, I tried to avoid the dance, but couldn't make that work, either. It's been a long, boring voyage :)

I did a new build last night, so I'll work it over today to make sure everything is ship shape on Darwin. It seems fine from poking around clang and GHC, but I'll update if I find a problem.

**Comment by acowley** [comment]
Everything looks good on Darwin. Can anyone help me navigate getting Linux and NixOS (and WIndows?) validation? I can wrangle VMs, but if there's existing infrastructure I'd rather use that.

**Comment by acowley** [comment]
I spun up a NixOS VM on EC2, set nixpkgs to a checkout of this branch, and did a `nixos-rebuild --switch` without any trouble. Installing `clang` also works fine. Not sure how much that evidence is worth. I'll try a non-NixOS linux when I have some time.

**Comment by acowley** [comment]
The last commit seems to get Linux up and running.

The trick there is that, as far as I can see, Linux was working because the grep executable came along with the bootstrap tools. Explicitly requiring `gnugrep` required a bit more plumbing.

Any help getting other stdenvs in order would be much appreciated!

**Comment by acowley** [comment]
There are rather a lot of arguably problematic shell scripts that make use of utilities. `write-mirror-list.sh` uses `grep`, and `setup.sh` from the generic stdenv uses `tar`, as examples.

Perhaps this kind of explicit dependency tracking isn't realistic for `nixpkgs`, and people would rather have these things white listed in some way. It takes hours of compiling to get to a point where these errors come out, and I'm not sure how much more of it I can reasonably do as this seems like perhaps a deliberate choice in the way `nixpkgs` is designed.

Non-Linux platforms could alternately be supported by patching on an as-needed basis.

**Comment by vcunat** [comment]
Staged after some minor fixups: 8f48a97 and 98c7e70.
